master/buildbot/steps/artifact.py

```diff
@@ -379,6 +448,108 @@ def start(self):
+        defer.returnValue(br)
+
+
+class DownloadArtifactFromParent(DownloadArtifact):
+    @defer.inlineCallbacks
+    def _getBuildRequest(self):
+        triggeredbybrid = self.build.requests[0].id
+        id = yield self.master.db.buildrequests.getTriggeredById(triggeredbybrid)
+        br = yield self.master.db.buildrequests.getBuildRequestById(id)
+        defer.returnValue(br)
+
+class DownloadArtifactFromChilden(LoggingBuildStep, CompositeStepMixin):
+    def __init__(self, name=None, projectPrefix=None, targetConfig=None, artifactBuilderName=None,
+                 artifact=None, artifactDirectory=None, artifactDestination=None,
+                 artifactServer=None, artifactServerDir=None, artifactServerPort=None,
+                 usePowerShell=True, baseLocalDir=None, **kwargs):
+        self.workdir = 'build'
+        self.artifactBuilderName = artifactBuilderName
+        self.artifact = artifact
+        self.projectPrefix = projectPrefix
+        self.target_config = targetConfig
+        self.artifactDirectory = artifactDirectory
+        self.artifactServer = artifactServer
+        self.artifactServerDir = artifactServerDir
+        self.artifactServerPort = artifactServerPort
+        self.artifactDestination = artifactDestination or artifact
+        self.master = None
+        self.usePowerShell = usePowerShell
+        self.kwargs = kwargs
+        self.name = name
+        self.baseLocalDir = baseLocalDir
+        LoggingBuildStep.__init__(self, **kwargs)
+
+    @defer.inlineCallbacks
+    def start(self):
+        if self.master is None:
+            self.master = self.build.builder.botmaster.parent
+
+        brid = self.build.requests[0].id
+        triggeredBuilderName = self.projectPrefix + self.target_config
+        partitionRequests = yield self.master.db.buildrequests.getBuildRequestsTriggeredBy(brid, triggeredBuilderName)
+        buildRequetsIdsWithArtifacts = self._getBuildRequestIdsWithArtifacts(partitionRequests)
+        self.partitionCount = len(buildRequetsIdsWithArtifacts)
+        self.stdio_log = self.addLogForRemoteCommands("stdio")
+        self.stdio_log.setTimestampsMode(self.timestamp_stdio)
+
+        for id in buildRequetsIdsWithArtifacts:
+            buildRequest = yield self.master.db.buildrequests.getBuildRequestById(id)
+            triggeredBuilderName = self.projectPrefix + self.target_config
+            artifactPath = "%s_%s_%s" % (safeTranslate(triggeredBuilderName),
+                                         id, FormatDatetime(buildRequest["submitted_at"]))
+
+            artifactPath += "/%s" % self.artifactDirectory
+            remotelocation = getRemoteLocation(self.artifactServer, self.artifactServerDir, artifactPath, "")
+            localdir = self.baseLocalDir + '/' + str(id)
+
+            command = [mkDir(self), '-p', '\''+ localdir + '\'']
+            yield self._docmd(command)
+
+            rsync = rsyncWithRetry(self, remotelocation, localdir, self.artifactServerPort)
```

Comment:
How are errors handled? - in my test I tried to download a non-existing artifact folder and got the following error:

`rsync: change_dir "/home/vagrant/artifacts/Simple_Test_661_02_07_2017_21_28_52_+0000/foobar" failed: No such file or directory (2)`

but the step still said `Success`